File: gunhoflash/merge-k-sorted-lists.py
from typing import Optional

class ListNode:
  def __init__(self, val = 0, next = None):
    self.val = val
    self.next = next

class Solution:
  def mergeTwoLists(self, list1: Optional[ListNode], list2: Optional[ListNode]) -> Optional[ListNode]:
    head = ListNode()
    tail = head

    while list1 and list2:
      if list1.val < list2.val:
        tail.next = list1
        tail = list1
        list1 = list1.next
      else:
        tail.next = list2
        tail = list2
        list2 = list2.next

    tail.next = list1 or list2

    return head.next

  def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
    if lists is None: return None

    length = len(lists)
    if length == 0: return None
    if length == 1: return lists[0]
    if length == 2: return self.mergeTwoLists(lists[0], lists[1])
    return self.mergeTwoLists(
      self.mergeKLists(lists[ : length >> 1]),
      self.mergeKLists(lists[length >> 1 : ])
    )

File: gunhoflash/merge-k-sorted-lists.ts
class ListNode {
  val: number;
  next: ListNode | null;
  constructor(val?: number, next?: ListNode | null) {
    this.val = val === undefined ? 0 : val;
    this.next = next === undefined ? null : next;
  }
}

function mergeTwoLists(
  list1: ListNode | null,
  list2: ListNode | null
): ListNode | null {
  const head = new ListNode();
  let tail = head;

  while (list1 && list2) {
    if (list1.val < list2.val) {
      tail = tail.next = list1;
      list1 = list1.next;
    } else {
      tail = tail.next = list2;
      list2 = list2.next;
    }
  }

  tail.next = list1 || list2;

  return head.next;
}

function mergeKLists(lists: Array<ListNode | null>): ListNode | null {
  if (lists === null) return null;

  const len = lists.length;
  if (len === 0) return null;
  if (len === 1) return lists[0];
  if (len === 2) return mergeTwoLists(lists[0], lists[1]);
  return mergeTwoLists(
    mergeKLists(lists.slice(0, len / 2)),
    mergeKLists(lists.slice(len / 2))
  );
}
